### PR TITLE
Fix dumping DLC to CIA.

### PR DESCRIPTION
**Pull request by Steveice10**
Instead of modifying the TMD content chunks, just remove missing contents from the CIA header's content index.

### CODE REVIEW

**Comment by d0k3** [comment]
Alright, I'll have a look tomorrow and then I'll merge it. Thanks for this!

**Comment by d0k3** [comment]
Okay... putting this up for discussion. There was a prompt earlier when DLCs are incomplete. Your pull request removed that prompt. Think we should leave it out? Given that DLCs are basically always incomplete, the prompt may be redundant. On the other hand, users may think that their DLCs are complete now (cause no prompt), which they aren't.

**Comment by Steveice10** [comment]
If you want the prompt to be there I can add it back. I just kind of figured users would understand it can only dump what they have.

**Comment by d0k3** [comment]
Nah, I'd rather just ask for your opinion. I think we can leave it as is.

That pull request is also tested?

**Comment by Steveice10** [comment]
Yes. I tested it by dumping all of my DLC, testing all of my DLC in citra with the DLC-fixing pull request, and reinstalling one of my game's DLC on my system using a dumped CIA.

File: arm9/source/utils/gameutil.c [hunk 1]
```diff
@@ -1227,24 +1227,23 @@ u32 BuildCiaFromTmdFileBuffered(const char* path_tmd, const char* path_cia, bool
...
-        memcpy(tmd->contentinfo[0].cmd_count, tmd->content_count, 2);
-        if (FixCiaHeaderForTmd(&(cia->header), tmd) != 0) return 1;
+        if ((fvx_stat(path_content, &fno) != FR_OK) || (fno.fsize != (u32) getbe64(chunk->size))) {
+            present[i] = false;
+
+            u16 index = getbe16(chunk->index);
+            cia->header.size_content -= getbe64(chunk->size);
+            cia->header.content_index[index/8] &= ~(1 << (7-(index%8)));
+        }
+        i++;
```
Comment (d0k3) [inline arm9/source/utils/gameutil.c]: This one can also be included in the `for` line now.

File: arm9/source/utils/gameutil.c [hunk 2]
```diff
@@ -1227,24 +1227,23 @@ u32 BuildCiaFromTmdFileBuffered(const char* path_tmd, const char* path_cia, bool
...
     if (dlc) for (u32 i = 0; (i < content_count) && (i < TMD_MAX_CONTENTS);) {
         FILINFO fno;
         TmdContentChunk* chunk = &(content_list[i]);
         snprintf(name_content, 256 - (name_content - path_content),
             (cdn) ? "%08lx" : (dlc && !cdn) ? "00000000/%08lx.app" : "%08lx.app", getbe32(chunk->id));
-        if ((fvx_stat(path_content, &fno) != FR_OK) || (fno.fsize != (u32) getbe64(chunk->size)))
-            memmove(chunk, chunk + 1, ((--content_count) - i) * sizeof(TmdContentChunk));
-        else i++;
-    }
-    if (!content_count) return 1;
```
Comment (d0k3) [inline arm9/source/utils/gameutil.c]: This check may still make sense and shouldn't be removed.

**Comment by d0k3** [comment]
I think at least the check should go back in. The other thing is nitpicking :^).

File: arm9/source/utils/gameutil.c (hunk 2 above)
Comment (Steveice10) [reply arm9/source/utils/gameutil.c]: The check also happens near the top of the function; the difference is now, content_count isn't modified, so it doesn't need to be repeated.

**Comment by Steveice10** [comment]
Comments fixed/responded to.

File: arm9/source/utils/gameutil.c [hunk 3]
```diff
@@ -1227,24 +1227,22 @@ u32 BuildCiaFromTmdFileBuffered(const char* path_tmd, const char* path_cia, bool
     if (!name_content) return 1; // will not happen
     name_content++;
     
+    bool present[TMD_MAX_CONTENTS];
+    memset(present, true, sizeof(present));
```
Comment (d0k3) [inline arm9/source/utils/gameutil.c]: Is `bool` guaranteed to be 1 byte here? Granted, this will work even if it's not.

Comment (d0k3) [reply arm9/source/utils/gameutil.c]: You're not sure, either, correct? Maybe `bool present[TMD_MAX_CONTENTS] = { true }` would be more sane?

Comment (Steveice10) [reply arm9/source/utils/gameutil.c]: It's implementation-defined, although I believe it is 1 byte in the case of devkitARM's environment.

Comment (Steveice10) [reply arm9/source/utils/gameutil.c]: If you want, I could flip it to "missing" and initialize it to all zeros instead of explicitly using "true".

Comment (d0k3) [reply arm9/source/utils/gameutil.c]: Okay, suggestion. Use an u8 array instead. Just to be safe. I know, this is again nitpicking, but I don't like the memset when we're not really sure about the size of the elements in the `present` array. Open for discussion.

**Comment by Steveice10** [comment]
Changed to u8 array.

**Comment by d0k3** [comment]
Merged, thanks! That was quick, also sorry about the nitpicking.

This also fixes #395

**Comment by jroweboy** [comment]
@d0k3 Sorry to pester you about this, but could we see a release with this fix soon? This will be required for DLC support in citra (for those people that don't all the DLC for the game) and when we announce that DLC is working, we'll be telling everyone to redump the dlc through godmode9 and that they should update godmode9 first :)

**Comment by d0k3** [comment]
@jroweboy - yeah, I intended to do one anyways. It will still take me a few days to wrap everything up and test ist.

**Pull request merged**